### DISCUSSION TITLE
Prefetching images from sanity to host on netlify

### DISCUSSION THREAD

**Discussion by procrates** [nuxt/nuxt#16639] [Questions] [answered]
Hi! I am building a static websites and one of the requirements is hosting everything on netlify including images. 

I am eagerly waiting for Nuxt Image to support predownloading images on static websites, but for now I really need to be able to populate the public folder with images from sanity. 

I have got it working for some images, but when there are several the images starts to corrupt.

I belive this is because the hook I am using 'build:before' is not waiting for the process to finish.

Anyone has an idea on how to solve this?

Here is the build hook I created:

```
'build:before': async () => {
      console.log('running build:before',)
      const images = await client.fetch('*[_type == "sanity.imageAsset"]')
      const imageBuilder = imageUrlBuilder(client)
      const imageWidths = [320, 640, 960, 1280, 1920]

      const imageUrls = images.flatMap((image) => {
        const imageSizes = imageWidths.map((width) => {
          return imageBuilder.image(image).auto('format').width(width).url()
        })
        return imageSizes
      })
      await handleImageDownload(imageUrls).then((res) => {
        console.log('res', res)
      })
        .catch((err) => {
          console.log('err :', err)
        })
    }
```
handleImageDownload downloads the images and saves them in public folder.

Thanks!

**Comment by procrates** [comment, marked as answer]
I am closing this question due to it not being general enough for the community. I will open a new one that is more general if i dont find the answer.

**Comment by TheAlexLichter** [comment]
@procrates Check out `<nuxt-img>` and their [static images](https://image.nuxtjs.org/getting-started/static) feature.

**Reply by procrates** [reply]
Thanks for the reply:) unfortunately this is for nuxt 2.

Nuxt 3 ain't ready yet but waiting patiently :D I know how much work they are doing on so many modules. 

[https://v1.image.nuxtjs.org/advanced/static-images](Static image Nuxt image v1)

I built a temporary solution for now using before build hook. It downloads and generates all images sizes, then I am just referencing those on frontend. Works pretty good so far. It is not a ton of images that needs to be generated.

Although maybe you have an answer on why hooks doesn't await the process to finish before going to the next step? It works on netlify, but not on dev.

**Reply by TheAlexLichter** [reply]
Ohh, you are right. My bad 🙈

**Reply by procrates** [reply]
Should have specified it is for Nuxt 3. But isnt this repo only for Nuxt 3?

**Reply by TheAlexLichter** [reply]
No no, you are more than right 😋 Mixed up the versions 😅

**Reply by procrates** [reply]
@manniL Do you know why the build hooks dont finish before the site has generated? I can see all the images are generated, but both on netlify and vercel the website is deployed before they are actually downloaded. 

And since nuxt 3 needs the images in the public directory before the build or generate this dosnt really work. I am really stuck here and on a timeline. Do you know how i would wait for the function to finish before building or generating the website?

I will update with a simple repro, but here is the code I am using to generate the images
```
hooks: {
    'build:before': async () => {
      if (process.env.NODE_ENV === 'development') return;
      const images = await client.fetch('*[_type == "sanity.imageAsset"]{extension, url, _id}')
      const imageBuilder = imageUrlBuilder(client)
      const imageWidths = [320, 640, 960, 1280, 1920]

      const imageUrls = await images.flatMap((image) => {
        if (image.extension === 'svg') {
          return {
            ext: image.extension,
            url: image.url,
            id: image._id,
            filePath: `${process.cwd()}/public/icons/`,
            fileName: `${image._id}.${image.extension}`
          }
        }
        const imageSizes = imageWidths.map((width) => {
          return {
            id: image._id,
            ext: image.extension,
            width,
            url: imageBuilder.image(image).width(width).url(),
            filePath: `${process.cwd()}/public/images/${width}/`,
            fileName: `${image._id}.${image.extension}`
          }
        })
        return imageSizes
      })
      console.log('img urls', imageUrls)
      const res = await Promise.all(
        [
          ...imageUrls
        ].map((imgUrl) => {
          const options = {
            url: imgUrl.url,
            dest: imgUrl.filePath,               // will be saved to /path/to/dest/image.jpg
          };
          return download.image(options)
            .then(({ filename }) => {
              console.log('Saved to', filename); // saved to /path/to/dest/image.jpg
            })
            .catch((err) => console.error(err));
        }))
      console.log('res', res)
    }
  },
```

**Reply by TheAlexLichter** [reply]
No spontaneous idea sadly - the hook is awaited properly 

https://github.com/nuxt/framework/blob/4f050e50bf2f17c315632d16589162e2865852c8/packages/nuxt3/src/core/builder.ts#L26

**Reply by procrates** [reply]
Thanks this is great! Looking to understand the framework indepth. Never really looked into the core code before, it is super interesting thou! 

What should i return in my hook for the await to continue? Because now it just continues without me returning anythingas you can see. Is there any reason for it to not finish?

**Reply by TheAlexLichter** [reply]
Ideally returning a promise. But having an `async` function should be enough so it will properly await it 🤔
